SideNavigation component

plbabin opened the pull request:
This update add a new SideNavigation component. The component is flexible enough to handle pretty much any situation that we have right now in HS-App. We should be able to recreate and improve all sidebar navigation.

The component can handle multiple state:

**The default one**
![screen recording 2019-02-21 at 05 09 pm](https://user-images.githubusercontent.com/203992/53205424-44929a80-35fc-11e9-80fd-964cda5b5703.gif)

**Collapsed**
![screen recording 2019-02-21 at 05 11 pm](https://user-images.githubusercontent.com/203992/53205431-478d8b00-35fc-11e9-9885-933917205918.gif)

**Floating menu**
![screen recording 2019-02-21 at 05 13 pm](https://user-images.githubusercontent.com/203992/53205433-49574e80-35fc-11e9-9636-594bb248729d.gif)

ItsJonQ commented:
@plbabin Merging in latest `master` for ya. It comes with Netlify and Travis performance boosts 👍

plbabin commented:
@jaredmcdaniel I added you as a design reviewer for this new UI component

tinkertrain commented on stories/SideNavigation.stories.js:
```diff
@@ -0,0 +1,324 @@
+  }
+}
+
+class SidebarFloatingMenu extends React.PureComponent {
```
On the floating feature.

It feels a little too manual to set it up (you have to position it yourself, remember to set up `position: relative` on the parent...) maybe it should be handled by the component itself?

a quick snippet:

```jsx
<SideNavigation
  floating
  floatingTriggerContent="🍔"
  isOpen={this.state.isOpen}
>
</SideNavigationFloatingUI>
```

Just a thought, there might be other things I'm not taking into consideration!

plbabin replied:
made a comment about that here: https://github.com/helpscout/hsds-react/pull/494#issuecomment-474898161

knicklabs commented on src/components/SideNavigation/SideNavigation.css.js:
```diff
@@ -0,0 +1,200 @@
+  border: `1px solid ${getColor('grey.500')}`,
+}
+
+export const SideNavigationUI = styled('div')`
```
+1 for using `nav` here.

plbabin replied:
updated in c4b173f1

knicklabs commented on src/components/SideNavigation/SideNavigation.css.js:
```diff
@@ -0,0 +1,200 @@
+  }
+`
+
+export const FooterUI = styled('div')`
```
+1

plbabin replied:
updated in c4b173f1

knicklabs commented on src/components/SideNavigation/SideNavigation.css.js:
```diff
@@ -0,0 +1,200 @@
+  }
+`
+
+export const HeaderUI = styled('div')`
```
+1

plbabin replied:
updated in c4b173f1

knicklabs commented on src/components/SideNavigation/Header.tsx:
```diff
@@ -0,0 +1,65 @@
+    if (collapsed) return <BadgeUI>{initial}</BadgeUI>
+
+    return (
+      <Heading size="h3" weight={500}>
```
For accessibility, this should have an `id`.

plbabin replied:
added in c4b173f1

knicklabs commented on src/components/SideNavigation/SideNavigation.tsx:
```diff
@@ -0,0 +1,95 @@
+    return (
+      <PropProvider value={this.getProviderValue()}>
+        <SideNavigationUI
+          aria-label="SideNavigation"
```
This should be `aria-labelledby` and refer to the `id` in the `Header` component.

plbabin replied:
thanks for pointing that out, added in c4b173f1

jaredmcdaniel commented:
@plbabin Looks like we're a ways out yet ... here's the list of things I noticed on first pass:

1) Title/Header font weight. Currently set to 500. It should be `400`.
2) I'm guessing the folder icons are just placeholders, but can we confirm we've got the right ones plugged in for the folders displayed?
3) `with dropdown header`
a) should have a caret to the right: http://c.hlp.sc/60e92e8df08f
b) It also has an active state (blue text): http://c.hlp.sc/a56c2286f854
c) Can we confirm the `li` active state in the dropdown? I'm not seeing it here.
4) Inactive folder counts are currently set to 400 font-weight. Should be `500`.
5) `with titled section`
a) can we add `-webkit-font-smoothing: antialiased;` to the section headers?
b) Section headers currently have 4px bottom padding. Can we make it `6px`?
6) `with regular button`
a) CTA should reflect the HSDS tertiary button styles: font weight 500, 30px tall, 15px horizontal padding: https://github.com/helpscout/hsds-product-ui/wiki/Buttons#tertiary
7) `with multiple section` Shouldn't this have a title? http://c.hlp.sc/d839d5d3d1e4
8) `with footer`
a) I know this is a factor of the icons themselves, but I'd just like to confirm that we can visually center these icons as-needed http://c.hlp.sc/076c794d68a8
b) The icon caret is pretty small: http://c.hlp.sc/45fd9f878dfd. Would like for it to be consistent with what's in production today: http://c.hlp.sc/b560d8ae530d
9) `is collapsed`
a) There should be no jumping around when the sidebar is expanded and collapsed. See production: http://c.hlp.sc/8d20447bad8c
b) In storybook here, when the sidebar is expanded is "pushes" the content section to the right. Want to make sure it "overlays" the content when expanded.
c) There should be a transition/animation when expanding/collapsing the sidebar vs just popping.
10) `is floating`
a) menu button toggle should be 36px x 36px and the menu icon is wrong http://c.hlp.sc/f647bc8d670c
b) When the sidebar is visible, the toggle button should have an active state http://c.hlp.sc/093550afc564
c) There should be no right pixel border on the sidebar, there should be a border radius on the top right and bottom right corners, and there should be a box shadow on the entire container: http://c.hlp.sc/10901c25ef5a
d) The icons for the "footer" action items are the wrong color: http://c.hlp.sc/35256d369b8c. Should be `grey-700` http://c.hlp.sc/375f29e1d7b3

plbabin commented:
thank you @jaredmcdaniel for that first pass. I should be able to tackle a lot of those points next monday

I think we should do a pass on Icons and import all svg that we have from our sketch file inside hsds-react. Or at least do the ones from the sidebar. Right now it's all placeholders, and it was to showcase the flexibility of the component

For all situation like `is collapsed` and `is floating`, I assume at first that some work would have been done inside hs-app to make it 100% perfect. Right now it's was just a demo, and not a reflection of the final result. I will follow Juan Pablo's suggestion and add some component to handle those situation inside hsds-react.

ItsJonQ commented:
@jaredmcdaniel Just a heads up, gonna be 🍐 'ing with @plbabin this afternoon :)

plbabin commented:
Just a heads-up that I will complete our React 16 PR first, and then circle back on this component early next week

ItsJonQ commented:
@plbabin How we lookin' for this one? Everything going okay with the React 16 upgrade?

plbabin commented:
I should be able to start back on that today

plbabin commented on src/components/SideNavigation/SideNavigation.css.js:
```diff
@@ -0,0 +1,200 @@
+  text-decoration: none;
+
+  &:hover {
+    color: ${getColor('charcoal.800')};
```
there is no hover state on prod right now. @jaredmcdaniel do you have a prefered hover color when the heading is a link (main mailbox page for example) ?

plbabin commented:
@jaredmcdaniel @tinkertrain After a quick discussion with @ItsJonQ, we decide to table the "is-floating" functionnality for this component. In a way everything that was done in this PR was "out-of-scope" for Customer List (even now we drop the sidebar component)

What I suggest is to circle back when we actually will implement the component in hs-app, and see how we want to build that hamburger menu pattern. We'll be able to extract the way it will be build into HSDS-react

hope that make sense!

I will update the PR to remove the story and documentation about the floating menu

jaredmcdaniel commented:
Coming along @plbabin! Diving right in:

1) Is there an example from the app where we use the `with linkable header` option? I'm just curious.
2) `with dropdown header`
a) Seems like the caret could be as large as 15px x 15px. Right now its pretty small at 12px.
b) Can we move the dropdown menu down 8px? Feels too tight with the heading right now
c) Possible to confirm what the `active` link styles are in the dropdown? Not seeing that here.
3) `Needs Attention` should have a background hover state consistent with all other links. No need for a color change on hover.
4) Can we confirm the default link color is `#556575`, (Charcoal, 400)
5) `Muted` links should have a text color change on hover consistent with other row hovers: http://c.hlp.sc/1e7a53d0df94
6) This new-conversation icon size still feels a little wonky but I think this is related to the overall icon sizing issue thats come up over the last couple days http://c.hlp.sc/c736ff445996
7) `with multiple sections`
a) Want to confirm the muted links have accurate hovers here as well. According to what I'm seeing for `is collapsed`, they do not.

plbabin commented:
> Is there an example from the app where we use the with linkable header option? I'm just curious.

@jaredmcdaniel I think the Mailbox heading is a link (https://github.com/helpscout/hsds-react/pull/494#discussion_r267097058)

plbabin commented:
> Coming along @plbabin! Diving right in:
> 
> 1. Is there an example from the app where we use the `with linkable header` option? I'm just curious.
> 2. `with dropdown header`
>    a) Seems like the caret could be as large as 15px x 15px. Right now its pretty small at 12px.
>    b) Can we move the dropdown menu down 8px? Feels too tight with the heading right now
>    c) Possible to confirm what the `active` link styles are in the dropdown? Not seeing that here.
> 3. `Needs Attention` should have a background hover state consistent with all other links. No need for a color change on hover.
> 4. Can we confirm the default link color is `#556575`, (Charcoal, 400)
> 5. `Muted` links should have a text color change on hover consistent with other row hovers: http://c.hlp.sc/1e7a53d0df94
> 6. This new-conversation icon size still feels a little wonky but I think this is related to the overall icon sizing issue thats come up over the last couple days http://c.hlp.sc/c736ff445996
> 7. `with multiple sections`
>    a) Want to confirm the muted links have accurate hovers here as well. According to what I'm seeing for `is collapsed`, they do not.

@jaredmcdaniel pushed a new update:

1. view last comment
2. a) caret, I added a 15 size to Icons, and apply that to the caret
b) done
c) Fixed a typo in all stories http://c.hlp.sc/9165bcaaf427. All dropdown should have a selected item now

3. seem to be a regression from something that was working. it’s all fixed now
4. I have charcoal.400 but it render with #4f5d6b
5. fixed
6. yes there is a scale applied on it, let’s circle back as soon as we are ok with a solution for that
7. I believe fixing 5 helped with that

jaredmcdaniel commented:
Following up on #1:

 I'm 99% sure the only time the header text should be a "link" is when it is a dropdown. It's never just a link that takes you to another page.

And I think I mentioned this in an earlier comment, but do we need a sidebar variation similar to the structure use for [tags](https://secure.helpscout.net/settings/tags/mailboxes//dateRange//start//end//sort/name/page/1)?

jaredmcdaniel commented:
Following up ... again ... on #1. I was 100% wrong, and the current styling and minimal hover styling is fine. 👍

plbabin commented:
> Following up on #1:
> 
> I'm 99% sure the only time the header text should be a "link" is when it is a dropdown. It's never just a link that takes you to another page.
> 
> And I think I mentioned this in an earlier comment, but do we need a sidebar variation similar to the structure use for [tags](https://secure.helpscout.net/settings/tags/mailboxes//dateRange//start//end//sort/name/page/1)?

we will need one. I wonder if it should be include in this PR since it's not really a component that is needed right away. We could go with what we have in place, and then improve the component when it's required by a HS-App 

thoughts?

knicklabs commented:
@plbabin I would not add additional functionality at this point. I think it would be best to card up that requirement on the HSDS Trello Board and we can grab it whenever is convenient.

plbabin commented:
@jaredmcdaniel could you do a last design review pass when you have a chance? I think we are in a very good place for a V1 of this component

thank you!

ItsJonQ commented:
@plbabin No rush! Just following up on this one 😊 

Anything I can do to help? Should we put this on hold for now?

ItsJonQ commented:
@knicklabs I agree! But first.. we need to fix tests (@plbabin )

:lolcry:

https://travis-ci.org/helpscout/hsds-react/builds/520376422?utm_source=github_status&utm_medium=notification

plbabin commented:
> @knicklabs I agree! But first.. we need to fix tests (@plbabin )
> 
> :lolcry:
> 
> https://travis-ci.org/helpscout/hsds-react/builds/520376422?utm_source=github_status&utm_medium=notification

@ItsJonQ last commit fixed all tests 😄

ItsJonQ commented:
@plbabin Awesome! Will merge + release once Travis is 💚

ItsJonQ commented:
@plbabin Lol. Tests are still breaking 🔥 😭

ItsJonQ commented:
Will take a looksy on my end

ItsJonQ commented:
@plbabin I think I got it. Not your fault... just, a weird thing that's happening with Emotion?

In the `SideNavigation.css.js` file, we're using nested selectors to reference `${Heading}`. This should work.. except the way that `Heading` is rendering is slightly unconventional, since you can change the HTML selector. 

The fix was just to use the `.c-Heading` className. Not ideal, but also not bad.

We still need to update tests to get coverage back up to 100%. The missing piece appears to be `SideNavigation/Button.tsx` for icon rendering, (line 44)

ItsJonQ commented:
Adding that test! Will merge/release when ready

plbabin commented:
thank you for fixing the latest test!
